sdk/core/Azure.Core/src/Pipeline/Internal/ResponseBodyPolicy.cs

```diff
@@ -68,50 +68,58 @@ private async ValueTask ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPip
             }
 
             Stream? responseContentStream = message.Response.ContentStream;
-            if (responseContentStream == null || responseContentStream.CanSeek)
+            if (responseContentStream == null)
             {
                 return;
             }
 
             if (message.BufferResponse)
             {
-                if (networkTimeout != Timeout.InfiniteTimeSpan)
+                // If the stream backing the response is already a MemoryStream which allows the underlying buffer to be exposed,
+                // we can reuse it. Otherwise, we need to copy the response into our own buffer.
+                //
+                // We require that the underlying buffer of the memory stream be publicly visible because Azure.Response.Content uses
+                // TryGetBuffer to fetch the underlying data (to wrap it in a BinaryData) without copying.
+                if (!IsMemoryStreamWithPubliclyVisibleBuffer(responseContentStream))
```

Comment:
I wonder if the fix should be here or in the `.Content` implementation itself. Both would result in a copy but if we do it in the property it's pay-for-play.

Reply:
Thought about that as well - the one concern I had was if we would replace the `ContentStream` on the Response as part of doing this (so future calls to `.Content` would not need to copy) and if doing so would be weird because of things like:

```
Stream s1 = response.ContentStream;
Console.WriteLine(response.Content);
Stream s2 = response.ContentStream;

Object.ReferenceEquals(s1, s2) // could return false?!
```

Do you think this would be surprising in practice?

Reply:
It's not exactly what I've meant. 

I was thinking of replacing 

https://github.com/Azure/azure-sdk-for-net/blob/0cecb0efd2435982e8cb0aa7b46a6dba6422b15f/sdk/core/Azure.Core/src/Response.cs#L75

with

```C#
return memoryContent.ToArray();
```

Reply:
Sounds reasonable to me, I suppose if we find that folks are calling `.Content` a bunch (and that's commonly leading to a bunch of copies) we can either figure a way to cache the value in Response (and eat the size increase) or explore something like swapping out the stream with a memory stream that exposes it's underlying buffer.